src/lin.py

```diff
@@ -1,16 +1,60 @@
 """Implementation of a linear time exact matching algorithm."""
 
 import argparse
+from asyncio.base_subprocess import ReadSubprocessPipeProto
+import numpy as np
 
+from utils import *
+
+
+
+def borderArray(p, x):
+    # Step 1: Construct s = p$x and its borderarray b
+    s = p + "$" + x
+    ba = np.zeros(len(s), dtype=int)
+    for i in range(1, len(s)):
+        b = ba[i - 1]
+        while b > 0 and s[i] != s[b]:
+            b = ba[b - 1]
+        if s[i] == s[b]:
+            ba[i] = b + 1
+        else:
+            ba[i] = 0
+    return ba, s
+
+
+def occurrence(p, x):
+    # Step 2: Report an occurence of p in x at position i-2m if b[i]=m
+    idx = []
+    ba = borderArray(p, x)[0]
+    s = borderArray(p, x)[1]
```

Comment:
You are computing the border array twice here, just so you can get both arrays out. That is entirely unnecessary, as you can do a parallel assignment instead:

```python
ba, s = borderArray(p, x)
```